clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -595,17 +609,38 @@ raw2trace_t::write_syscall_template(raw2trace_thread_data_t *tdata, int syscall_
                        "template.";
         return false;
     }
-    trace_entry_t *buf_base = get_write_buffer(tdata);
-    trace_entry_t *buf = buf_base;
+    // Write out delayed branches if any. When we're at a system call number marker
+    // any delayed branches should already be written out at the prior syscall
+    // instruction. However, on Win32 the WOW64 calls are treated as a system call
+    // (has a corresponding system call number marker) but is also a CTI so is
+    // delayed. We expect delayed branches only in that case.
+    if (delayed_branches_exist(tdata)) {
+        app_pc next_pc = get_first_app_pc_for_syscall_template(syscall_num);
+        if (next_pc == nullptr) {
+            tdata->error = "Could not find first app pc in system call template for " +
+                std::to_string(syscall_num);
+            return false;
+        }
+        if (!append_delayed_branch(tdata, next_pc))
+            return false;
+    }
+    trace_entry_t *buf = reinterpret_cast<trace_entry_t *>(buf_in);
     trace_entry_t start_entry = { TRACE_TYPE_MARKER,
                                   TRACE_MARKER_TYPE_SYSCALL_TRACE_START,
                                   { static_cast<addr_t>(syscall_num) } };
     *buf = start_entry;
     ++buf;
+    // Now write any accumulated entries from before and the start entry.
```

Comment:
s/ and/, plus/ would make this easier to read

Reply:
Done.